# Clase C3 - Numpy

## Temario

- Que es numpy (presentacion)
- ND-Arrays (presentacion)
- Indexing (Ejemplo con presentacion)
- Bradcasting
- Cambiar tipo de datos
- Funciones especificas
    - np.genfromtxt()
    - cumprod()
    - enumerate()
    - np.reshape()
    - np.diff()

## Temario extra practica

- Metodos de inicializacion de arrays
- Random (lo vamos a retomar proxima clase)
- Eficiencia de numpy

## Organizacion prevista para la clase de hoy

- Consultas y puesta en comun de dudas o repaso de los temas de la clase pasada (15mins)
- Presentacion de Numpy con ejemplo (45mins)
- Hands-On (15mins)
- Ejemplo mas completo (1h)




# Presentación

# Hands-On

Ejercicios extra:

- Hacer un codigo que arme los primeros m elementos las tablas de multiplicar de un numero arbitrario N usando broadcasting y np.narray()
    - Leer la documentacion: https://numpy.org/doc/stable/reference/generated/numpy.arange.html

- Hacer una funcion que acepte un entero o una lista de enteros y arme sus tablas de multiplicar hasta el 10
    - Leer la documentacion de reshape https://numpy.org/doc/stable/reference/generated/numpy.reshape.html y https://stackoverflow.com/questions/18691084/what-does-1-mean-in-numpy-reshape

In [36]:
# Soluciones posibles

import numpy as np

m = 10
N = 7
escala = np.arange(1,m+1)
multiplos = escala * N
display (multiplos)

def tablas_op1(numeros: list):
    if not (type(numeros) == list or type(numeros) == int):
        raise ValueError(f'Se espera que el parametro numeros sea un entero o una lista de enteros. El valor "{numeros}" no cumple esa condición.')
    if type(numeros) == list:
        for elemento in numeros:
            if type(elemento) != int:
                raise ValueError(f'Se espera que el parametro numeros sea un entero o una lista de enteros. El valor "{elemento}" no cumple esa condición.')
    numeros = np.array(numeros)
    multiplos = np.arange(1,11).reshape(-1,1)
    return numeros*multiplos

def tablas_op2(numeros: list):
    if type(numeros) == int:
        numeros = [numeros]
    if not type(numeros) == list:
        raise ValueError(f'Se espera que el parametro numeros sea un entero o una lista de enteros. El valor "{numeros}" no cumple esa condición.')
    for elemento in numeros:
        if type(elemento) != int:
            raise ValueError(f'Se espera que el parametro numeros sea un entero o una lista de enteros. El valor "{elemento}" no cumple esa condición.')
    numeros = np.array(numeros)
    multiplos = np.arange(1,11).reshape(-1,1)
    return numeros*multiplos

display (tablas_op1(1))
display (tablas_op2(1.5))

try:
    display (tablas_op1(1))
except:
    print ('Hay un error en la funcion que se utilizo, pero no nos importa, no sabemos cual es y seguimos adelante como si nada.')

print ('Fin del código.')

array([ 7, 14, 21, 28, 35, 42, 49, 56, 63, 70])

array([[ 1],
       [ 2],
       [ 3],
       [ 4],
       [ 5],
       [ 6],
       [ 7],
       [ 8],
       [ 9],
       [10]])

ValueError: Se espera que el parametro numeros sea un entero o una lista de enteros. El valor "1.5" no cumple esa condición.

# Problema de ejemplo de uso

Tenemos datos de la inflación mes a mes en termino porcentuales. Y tenemos la serie de precios correspondientes a cada mes de 6 productos. Queremos saber:

- Si cada producto aumento su precio mas, igual o menos que la inflacion tomando los doce meses en su conjunto.
- En que meses que productos aumentaron mas que la inflacion o menos. 
- Durante que periodos los productos tuvieron sus precios por debajo de lo que corresponderia segun la inflacion y cuales por encima. 
- Lo mismo para una canasta compuesta por: 10 producto 1, 1 producto 2, 500 productos 3, 5 productos 4, 0.25 producto 5 y 1200 productos 6 (este queda de ejecicio)

![datos del problema](datos.png)

In [82]:
# Lo primero que vamos a hacer es cargar los datos. No vamos a usar pandas porque hoy queremos limitarnos a numpy

import numpy as np

In [83]:
filename = "./C3-Ejemplo.csv"

data_cruda = np.genfromtxt(filename, delimiter=',', dtype = str) # Notar que si no ponemos el separador correcto tira error. Y si no ponemos dtype lee todo como vacio.

data_cruda

array([['', '', '', '', '', '', '', '', '', '', '', '', ''],
       ['Mes', 'Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio',
        'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre',
        'Diciembre'],
       ['Inflacion mensual (%)', '1', '1.5', '3', '2.3', '2', '1.9',
        '1.8', '2.5', '1', '0.5', '1', '1.1'],
       ['Producto 1', '123', '124.845', '129.8388', '132.8250924',
        '135.4815942', '139.4105605', '144.7081618', '148.3258658',
        '152.7756418', '153.53952', '158.1457056', '163.0482225'],
       ['Producto 2', '5563', '5646.445', '5759.3739', '5891.8395',
        '5950.757895', '6004.314716', '6172.435528', '6265.022061',
        '6327.672281', '6296.03392', '6296.03392', '6428.250632'],
       ['Producto 3', '12.4', '12.586', '12.83772', '13.00461036',
        '13.13465646', '13.25286837', '13.49142', '13.8287055',
        '13.8287055', '13.89784903', '13.89784903', '14.05072537'],
       ['Producto 4', '108', '106.38', '104.2524', '111.862825

In [84]:
# Mirando los datos vamos a separar la parte de etiquetas que son texto de los numeros. Y los numeros los vamos a transformar a ese tipo.

etiqueta_meses = data_cruda[1,1:]
display (etiqueta_meses)

etiqueta_productos = data_cruda[3:,0]
display (etiqueta_productos)

serie_inflacion_mensual = data_cruda[2,1:].astype(float) # Notar como se ve si no ponemos as float
display (serie_inflacion_mensual)

precios = data_cruda[3:,1:].astype(float)
display (precios)

# Con https://numpy.org/doc/stable/reference/generated/numpy.set_printoptions.html podemos configurar como se imprime. Vamos a sacarle el punto flotante.

np.set_printoptions(suppress=True)

display (precios)

array(['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio',
       'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre'],
      dtype='<U21')

array(['Producto 1', 'Producto 2', 'Producto 3', 'Producto 4',
       'Pruducto 5', 'Producto 6'], dtype='<U21')

array([1. , 1.5, 3. , 2.3, 2. , 1.9, 1.8, 2.5, 1. , 0.5, 1. , 1.1])

array([[  123.        ,   124.845     ,   129.8388    ,   132.8250924 ,
          135.4815942 ,   139.4105605 ,   144.7081618 ,   148.3258658 ,
          152.7756418 ,   153.53952   ,   158.1457056 ,   163.0482225 ],
       [ 5563.        ,  5646.445     ,  5759.3739    ,  5891.8395    ,
         5950.757895  ,  6004.314716  ,  6172.435528  ,  6265.022061  ,
         6327.672281  ,  6296.03392   ,  6296.03392   ,  6428.250632  ],
       [   12.4       ,    12.586     ,    12.83772   ,    13.00461036,
           13.13465646,    13.25286837,    13.49142   ,    13.8287055 ,
           13.8287055 ,    13.89784903,    13.89784903,    14.05072537],
       [  108.        ,   106.38      ,   104.2524    ,   111.8628252 ,
          104.0324274 ,   102.8880707 ,   115.0288631 ,   121.3554506 ,
          120.141896  ,   114.7355107 ,   125.0617067 ,   117.683066  ],
       [13234.        , 14094.21      , 14376.0942    , 15569.31002   ,
        16347.77552   , 16331.42774   , 16625.39344   , 1770

array([[  123.        ,   124.845     ,   129.8388    ,   132.8250924 ,
          135.4815942 ,   139.4105605 ,   144.7081618 ,   148.3258658 ,
          152.7756418 ,   153.53952   ,   158.1457056 ,   163.0482225 ],
       [ 5563.        ,  5646.445     ,  5759.3739    ,  5891.8395    ,
         5950.757895  ,  6004.314716  ,  6172.435528  ,  6265.022061  ,
         6327.672281  ,  6296.03392   ,  6296.03392   ,  6428.250632  ],
       [   12.4       ,    12.586     ,    12.83772   ,    13.00461036,
           13.13465646,    13.25286837,    13.49142   ,    13.8287055 ,
           13.8287055 ,    13.89784903,    13.89784903,    14.05072537],
       [  108.        ,   106.38      ,   104.2524    ,   111.8628252 ,
          104.0324274 ,   102.8880707 ,   115.0288631 ,   121.3554506 ,
          120.141896  ,   114.7355107 ,   125.0617067 ,   117.683066  ],
       [13234.        , 14094.21      , 14376.0942    , 15569.31002   ,
        16347.77552   , 16331.42774   , 16625.39344   , 1770

# Pensar la matematica que necesitamos resolver

Tenemos la inflacion mes a mes en termino porcentuales. Si en un periodo hay 5% de inflacion y el precio original de algo es X entonces al finalizar el periodo la cuenta que debemos hacer es $X_f = X_i \cdot (\frac{5}{100} + 1 )$ Y debemos repetir esa cuenta tomando en cada periodo el resultado del periodo anterior.

Veamos como hacer esa transformacion con y cuenta de forma piola con numpy

In [85]:
inflacion_multiplicativa = serie_inflacion_mensual/100 + 1 # Aca entra en juego el tema de broadcasting. Como dividimos un vector por un numero divide elemento a elemento (esto no funciona con listas), y lo mismo al sumar un numero
display (inflacion_multiplicativa)

# De esta serie vamos a formzar el primer valor porque es el valor de referencia y no tenemos valor de los productos antes de ese periodo. Tiene sentido que el precio relativo al iniciar la serie sea 1

inflacion_multiplicativa[0] = 1
display (inflacion_multiplicativa)

# Ahora queremos que multiplique un numero arbitrario (que vamos a tomar como 1 que es lo mismo que no ponerlo) primero por 1.01 y a ese resultado por 1.015 y asi sucesivamente

# Podemos implementarlo a mano:
valor = 1
for factor in inflacion_multiplicativa:
    valor = valor * factor
    print (valor)

# Pero numpy tiene eso resuelto

inflacion_acumulada = inflacion_multiplicativa.cumprod()
display (inflacion_acumulada)

# Contra estos valores tenemos que comparar la variacion de precio de los productos.

array([1.01 , 1.015, 1.03 , 1.023, 1.02 , 1.019, 1.018, 1.025, 1.01 ,
       1.005, 1.01 , 1.011])

array([1.   , 1.015, 1.03 , 1.023, 1.02 , 1.019, 1.018, 1.025, 1.01 ,
       1.005, 1.01 , 1.011])

1.0
1.015
1.04545
1.06949535
1.090885257
1.111612076883
1.131621094266894
1.1599116216235663
1.171510737839802
1.177368291529001
1.189141974444291
1.2022225361631782


array([1.        , 1.015     , 1.04545   , 1.06949535, 1.09088526,
       1.11161208, 1.13162109, 1.15991162, 1.17151074, 1.17736829,
       1.18914197, 1.20222254])

## Como responder la primer pregunta

Queremos saber cual hubiese sido el precio del producto 1 si hubiese incrementado su precio segun la inflacion.

Para eso el primer mes tenemos que multiplicar el precio 123 por 1,01 el segundo por 1.02515 y asi... Con esa cuenta tenemos cuanto deberia costar el producto cada mes si siguiera la inflacion, y despues hay que comparar con la variacion real del precio.


In [86]:
precio_prod1 = precios[0]
#display (precio_prod1)
precio_teorico_prod1 = precios[0,0] * inflacion_acumulada

print (precio_teorico_prod1)

# Si queremos saber la inflacion acumulada en todo el periodo lo que hay que hacer es dividir el precio final de las cosas sobre el del primer mes

print (f'La inflacion acumulada de todo el periodo segun el indice de inflación es {(inflacion_acumulada[-1] - 1) * 100}%') # Aca no dividimos por uno porque es redundante
print (f'La inflacion acumulada de todo el periodo para el producto 1 es {(precio_prod1[-1]/precio_prod1[0] - 1) * 100}%') 

# Vemos que el producto 1 subio por encima de la inflacion. 

[123.         124.845      128.59035    131.54792805 134.17888661
 136.72828546 139.18939459 142.66912946 144.09582075 144.81629986
 146.26446286 147.87337195]
La inflacion acumulada de todo el periodo segun el indice de inflación es 20.222253616317822%
La inflacion acumulada de todo el periodo para el producto 1 es 32.55953048780489%


Ahora automaticemos esta logica en una serie de cuentas que lo resuelva en unos pocos pasos para todos los productos.

In [87]:
valor_referencia = (inflacion_acumulada[-1] - 1) * 100
print (f'La inflacion acumulada de todo el periodo segun el indice de inflación es {valor_referencia:.2f}%') # Aplicamos formato para que muestre dos decimales
# Buscamos el indice para el cociente de los precios reales
# display (precios[:,-1]/precios[:,0])
for nombre, aumento in zip(etiqueta_productos, precios[:,-1]/precios[:,0]):
    #print (f'La inflacion acumulada de todo el periodo para {nombre} es {(aumento - 1) * 100:.2f}%') 
    if aumento > inflacion_acumulada[-1]:
        print (f'El producto {nombre} aumento por encima de la inflación')
    else:
        print (f'El producto {nombre} aumento por debajo de la inflación')


La inflacion acumulada de todo el periodo segun el indice de inflación es 20.22%
El producto Producto 1 aumento por encima de la inflación
El producto Producto 2 aumento por debajo de la inflación
El producto Producto 3 aumento por debajo de la inflación
El producto Producto 4 aumento por debajo de la inflación
El producto Pruducto 5 aumento por encima de la inflación
El producto Producto 6 aumento por debajo de la inflación


Vamos ahora a responder la pregunta 3, es decir, queremos saber que meses los precios de los productos obervados se mantienen por encima o por debajo de lo que deberian valor si hubiesen acompañado a la inflación.

Para eso tenemos que tomar el precio inicial de cada producto, calcular cuanto deberian valer mes a mes segun la inflacion y comparar con el valor real. Comparar esos dos valores e informar en que meses fue menor y cuales mayor.

In [88]:
# Para empezar queremos tomar los precios iniciales de cada producto y multiplicar cada uno de ellos por toda la serie de inflacion. 
# Veamos las dimencionalidades de los elementos que queremos multiplicar
display (precios[:,0].shape)
display (inflacion_acumulada.shape)

# No sirve, queremos que uno sea de 1x6 y el otro de 12x1 para que nos quede una cosa de 12x6 que es el precio los doce meses de los 6 productos

# https://stackoverflow.com/questions/36384760/transforming-a-row-vector-into-a-column-vector-in-numpy

display (precios[:,0])
display (precios[:,0].reshape(-1,1))
precios_teoricos = precios[:,0].reshape(-1,1) * inflacion_acumulada
display (precios_teoricos)

(6,)

(12,)

array([  123. ,  5563. ,    12.4,   108. , 13234. ,     3. ])

array([[  123. ],
       [ 5563. ],
       [   12.4],
       [  108. ],
       [13234. ],
       [    3. ]])

array([[  123.        ,   124.845     ,   128.59035   ,   131.54792805,
          134.17888661,   136.72828546,   139.18939459,   142.66912946,
          144.09582075,   144.81629986,   146.26446286,   147.87337195],
       [ 5563.        ,  5646.445     ,  5815.83835   ,  5949.60263205,
         6068.59468469,  6183.8979837 ,  6295.20814741,  6452.58835109,
         6517.1142346 ,  6549.69980578,  6615.19680383,  6687.96396868],
       [   12.4       ,    12.586     ,    12.96358   ,    13.26174234,
           13.52697719,    13.78398975,    14.03210157,    14.38290411,
           14.52673315,    14.59936681,    14.74536048,    14.90755945],
       [  108.        ,   109.62      ,   112.9086    ,   115.5054978 ,
          117.81560776,   120.0541043 ,   122.21507818,   125.27045514,
          126.52315969,   127.15577549,   128.42733324,   129.84003391],
       [13234.        , 13432.51      , 13835.4853    , 14153.7014619 ,
        14436.77549114, 14711.07422547, 14975.87356153, 1535

In [89]:
# Ahora vamos a comparar

precio_producto_atrasado = precios_teoricos < precios

display (precio_producto_atrasado)

# Tambien podriamos ver el nivel de atraso

atraso_productos = precios_teoricos / precios

display (atraso_productos)


# Notar que el 1 es engañoso!!! Eso es porque si hay una diferencia muy sutil puede haber cuestiones de redondeo!

array([[False,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True],
       [False, False, False, False, False, False, False, False, False,
        False, False, False],
       [False,  True, False, False, False, False, False, False, False,
        False, False, False],
       [False, False, False, False, False, False, False, False, False,
        False, False, False],
       [False,  True,  True,  True,  True,  True,  True,  True,  True,
         True,  True,  True],
       [False, False, False, False, False, False, False, False, False,
        False, False, False]])

array([[1.        , 1.        , 0.99038462, 0.99038462, 0.99038462,
        0.98075989, 0.96186278, 0.96186278, 0.94318583, 0.94318583,
        0.92487154, 0.90693029],
       [1.        , 1.        , 1.00980392, 1.00980392, 1.01980198,
        1.02990904, 1.01989047, 1.02993865, 1.02993865, 1.04028979,
        1.05069269, 1.04040187],
       [1.        , 1.        , 1.00980392, 1.01977237, 1.02986913,
        1.04007596, 1.04007596, 1.04007596, 1.05047672, 1.05047672,
        1.06098148, 1.06098148],
       [1.        , 1.03045685, 1.08303118, 1.03256375, 1.13248927,
        1.16684183, 1.06247315, 1.03226064, 1.05311439, 1.10825127,
        1.02691173, 1.10330261],
       [1.        , 0.95305164, 0.96239529, 0.90907699, 0.88310336,
        0.90078311, 0.90078311, 0.86695088, 0.88446504, 0.81925102,
        0.76615142, 0.75128912],
       [1.        , 1.        , 1.03      , 1.02002904, 1.05093901,
        1.04072581, 1.02067329, 1.00113887, 1.03178598, 1.06353324,
        1.07416857,

In [90]:
# Vamos a reportar los meses en que hubo atraso para cada producto

for i, nombre in enumerate(etiqueta_productos):
    #print (i)
    mascara = precio_producto_atrasado[i]
    #print (etiqueta_meses[mascara])
    for mes in etiqueta_meses[mascara]:
        print (f'Tomando como referencia el mes de enero, el mes {mes} el producto {nombre} mantuvo su precio retrasado respecto a la inflación')

Tomando como referencia el mes de enero, el mes Febrero el producto Producto 1 mantuvo su precio retrasado respecto a la inflación
Tomando como referencia el mes de enero, el mes Marzo el producto Producto 1 mantuvo su precio retrasado respecto a la inflación
Tomando como referencia el mes de enero, el mes Abril el producto Producto 1 mantuvo su precio retrasado respecto a la inflación
Tomando como referencia el mes de enero, el mes Mayo el producto Producto 1 mantuvo su precio retrasado respecto a la inflación
Tomando como referencia el mes de enero, el mes Junio el producto Producto 1 mantuvo su precio retrasado respecto a la inflación
Tomando como referencia el mes de enero, el mes Julio el producto Producto 1 mantuvo su precio retrasado respecto a la inflación
Tomando como referencia el mes de enero, el mes Agosto el producto Producto 1 mantuvo su precio retrasado respecto a la inflación
Tomando como referencia el mes de enero, el mes Septiembre el producto Producto 1 mantuvo su pr

Veamos ahora como responder la segunda pregunta donde lo que hay que resolver si ese mes el producto aumento respecto al mes anterior mas o menos que la inflacion (que no es lo mismo que tomar enero por referencia y ver la acumulación desde enero)

Para eso hay que ver cual fue la variacion intermensual y dividir respecto al mes justo anterior (y sumar 1). No es la unica manera de hacerlo pero lo vamos a hacer asi 

In [91]:
variaciones_precio = np.diff(precios)
display (variaciones_precio) # Vemos que por ejemplo algunos meses bajo el precio
porcentaje_aumento = variaciones_precio/precios[:,:-1]
display (porcentaje_aumento)
porcentaje_aumento = porcentaje_aumento * 100
display (porcentaje_aumento)
# Ahora comparamos con el aumento de la inflacion 
porcentaje_aumento_relativo = (porcentaje_aumento) / serie_inflacion_mensual[1:]
display (porcentaje_aumento_relativo)

array([[   1.845     ,    4.9938    ,    2.9862924 ,    2.6565018 ,
           3.9289663 ,    5.2976013 ,    3.617704  ,    4.449776  ,
           0.7638782 ,    4.6061856 ,    4.9025169 ],
       [  83.445     ,  112.9289    ,  132.4656    ,   58.918395  ,
          53.556821  ,  168.120812  ,   92.586533  ,   62.65022   ,
         -31.638361  ,    0.        ,  132.216712  ],
       [   0.186     ,    0.25172   ,    0.16689036,    0.1300461 ,
           0.11821191,    0.23855163,    0.3372855 ,    0.        ,
           0.06914353,    0.        ,    0.15287634],
       [  -1.62      ,   -2.1276    ,    7.6104252 ,   -7.8303978 ,
          -1.1443567 ,   12.1407924 ,    6.3265875 ,   -1.2135546 ,
          -5.4063853 ,   10.326196  ,   -7.3786407 ],
       [ 860.21      ,  281.8842    , 1193.21582   ,  778.4655    ,
         -16.34778   ,  293.9657    , 1080.65058   , -177.06044   ,
        1489.9636    , 1521.51578   ,  636.75435   ],
       [   0.045     ,    0.        ,    0.100485 

array([[ 0.015,  0.04 ,  0.023,  0.02 ,  0.029,  0.038,  0.025,  0.03 ,
         0.005,  0.03 ,  0.031],
       [ 0.015,  0.02 ,  0.023,  0.01 ,  0.009,  0.028,  0.015,  0.01 ,
        -0.005,  0.   ,  0.021],
       [ 0.015,  0.02 ,  0.013,  0.01 ,  0.009,  0.018,  0.025,  0.   ,
         0.005,  0.   ,  0.011],
       [-0.015, -0.02 ,  0.073, -0.07 , -0.011,  0.118,  0.055, -0.01 ,
        -0.045,  0.09 , -0.059],
       [ 0.065,  0.02 ,  0.083,  0.05 , -0.001,  0.018,  0.065, -0.01 ,
         0.085,  0.08 ,  0.031],
       [ 0.015,  0.   ,  0.033, -0.01 ,  0.029,  0.038,  0.045, -0.02 ,
        -0.025,  0.   ,  0.011]])

array([[ 1.5       ,  4.        ,  2.3       ,  1.99999996,  2.90000005,
         3.8       ,  2.49999997,  3.00000002,  0.49999999,  3.        ,
         3.10000002],
       [ 1.5       ,  2.        ,  2.30000001,  1.        ,  0.9       ,
         2.8       ,  1.5       ,  0.99999999, -0.49999999,  0.        ,
         2.09999999],
       [ 1.5       ,  2.        ,  1.3       ,  0.99999997,  0.90000001,
         1.8       ,  2.5       ,  0.        ,  0.50000002,  0.        ,
         1.1       ],
       [-1.5       , -2.        ,  7.3       , -7.00000003, -1.1       ,
        11.80000006,  5.50000003, -1.00000008, -4.49999998,  9.00000003,
        -5.9       ],
       [ 6.5       ,  2.        ,  8.30000001,  4.99999999, -0.10000003,
         1.8       ,  6.50000004, -1.        ,  8.49999998,  8.00000003,
         3.09999999],
       [ 1.5       ,  0.        ,  3.3       , -1.        ,  2.89999999,
         3.8       ,  4.50000002, -2.00000002, -2.5       ,  0.        ,
         1.1  

array([[ 1.        ,  1.33333333,  1.        ,  0.99999998,  1.52631582,
         2.11111111,  0.99999999,  3.00000002,  0.99999999,  3.        ,
         2.81818183],
       [ 1.        ,  0.66666667,  1.        ,  0.5       ,  0.47368421,
         1.55555556,  0.6       ,  0.99999999, -0.99999999,  0.        ,
         1.9090909 ],
       [ 1.        ,  0.66666667,  0.56521739,  0.49999999,  0.47368422,
         1.        ,  1.        ,  0.        ,  1.00000004,  0.        ,
         1.        ],
       [-1.        , -0.66666667,  3.17391304, -3.50000002, -0.57894737,
         6.55555559,  2.20000001, -1.00000008, -8.99999997,  9.00000003,
        -5.36363637],
       [ 4.33333333,  0.66666667,  3.60869566,  2.5       , -0.05263159,
         1.        ,  2.60000002, -1.        , 16.99999995,  8.00000003,
         2.81818181],
       [ 1.        ,  0.        ,  1.43478261, -0.5       ,  1.52631578,
         2.11111111,  1.80000001, -2.00000002, -4.99999999,  0.        ,
         1.   

In [92]:
# Ahora hacemos el reporte

for i, mes in enumerate(etiqueta_meses[1:]): # a proposito lo recorro diferente para practicar. Probar porque ponemos 1:
    mascara_productos_aumento_excesivo = porcentaje_aumento_relativo[:,i] > 1
    # print (mascara_productos_aumento_excesivo)
    print (f'La lista de productos que aumentaron mas que la inflación en el mes de {mes} son: {etiqueta_productos[mascara_productos_aumento_excesivo]}')


La lista de productos que aumentaron mas que la inflación en el mes de Febrero son: ['Pruducto 5']
La lista de productos que aumentaron mas que la inflación en el mes de Marzo son: ['Producto 1']
La lista de productos que aumentaron mas que la inflación en el mes de Abril son: ['Producto 2' 'Producto 4' 'Pruducto 5' 'Producto 6']
La lista de productos que aumentaron mas que la inflación en el mes de Mayo son: ['Pruducto 5']
La lista de productos que aumentaron mas que la inflación en el mes de Junio son: ['Producto 1' 'Producto 6']
La lista de productos que aumentaron mas que la inflación en el mes de Julio son: ['Producto 1' 'Producto 2' 'Producto 4' 'Pruducto 5' 'Producto 6']
La lista de productos que aumentaron mas que la inflación en el mes de Agosto son: ['Producto 3' 'Producto 4' 'Pruducto 5' 'Producto 6']
La lista de productos que aumentaron mas que la inflación en el mes de Septiembre son: ['Producto 1']
La lista de productos que aumentaron mas que la inflación en el mes de Oct